### ISSUE TITLE
relay.py and relaymanager.py should conform to docstring and whitespace standards

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#6739]
|[<img alt="staceysern's avatar" src="https://avatars.githubusercontent.com/u/3631162?s=50" width="50" height="50">](https://github.com/staceysern)| @staceysern reported|
|-|-|
|Trac ID|trac#6739|
|Type|enhancement|
|Created|2013-09-12 01:41:55Z|
|Branch|https://github.com/twisted/twisted/tree/relay-doc-6739|

The docstrings in relay.py and relaymanager.py should be updated to provide descriptions of all classes, functions and methods as well as their parameters and attributes. The spacing before classes and functions should conform to the coding standards.

Some other tickets involve changing code in these files so some docstrings will be changed under those tickets instead of this one.  The changes omitted from this ticket are:

relay.py
 * [#6383](https://github.com/twisted/twisted/issues/6383): DomainQueuer, DomainQueuer.__init__ and DomainQueuer.willRelay
 * [[#6717](https://github.com/twisted/twisted/issues/6717)](https://github.com/twisted/twisted/issues/6717): All docstrings for RelayerMixin

relaymanager.py
 * #6717: All docstrings for ManagedRelayerMixin
 * [#6719](https://github.com/twisted/twisted/issues/6719): SmartHostSMTPRelayingManager._checkStateMX
 * [#6705](https://github.com/twisted/twisted/issues/6705): All docstrings for _AttemptManager

<details><summary>Searchable metadata</summary>

```
trac-id__6739 6739
type__enhancement enhancement
reporter__shira shira
priority__normal normal
milestone__ 
branch__branches_relay_doc_6739 branches/relay-doc-6739
branch_author__shira shira
status__closed closed
resolution__fixed fixed
component__mail mail
keywords__documentation documentation
time__1378950115000000 1378950115000000
changetime__1389544398000000 1389544398000000
version__None None
owner__shira shira

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="staceysern's avatar" src="https://avatars.githubusercontent.com/u/3631162?s=50" width="50" height="50">](https://github.com/staceysern)<a name="note_1"></a>|@staceysern commented|
|-|-|

(In [39958]) Branching to 'relay-doc-6739'

**Comment by twisted-trac** [comment]
|[<img alt="staceysern's avatar" src="https://avatars.githubusercontent.com/u/3631162?s=50" width="50" height="50">](https://github.com/staceysern)<a name="note_2"></a>|@staceysern commented|
|-|-|

The list above should read:

 
relay.py 
 * [#6383](https://github.com/twisted/twisted/issues/6383): DomainQueuer, DomainQueuer.init and DomainQueuer.willRelay 
 * [[#6719](https://github.com/twisted/twisted/issues/6719)](https://github.com/twisted/twisted/issues/6719): All docstrings for RelayerMixin

relaymanager.py
 * #6719: All docstrings for ManagedRelayerMixin
 * [#6717](https://github.com/twisted/twisted/issues/6717): SmartHostSMTPRelayingManager._checkStateMX
 * [#6705](https://github.com/twisted/twisted/issues/6705): All docstrings for _AttemptManager

**Comment by twisted-trac** [comment]
|[<img alt="staceysern's avatar" src="https://avatars.githubusercontent.com/u/3631162?s=50" width="50" height="50">](https://github.com/staceysern)<a name="note_3"></a>|@staceysern commented|
|-|-|

(In [40017]) Revise docstrings and whitespace

refs [#6739](https://github.com/twisted/twisted/issues/6739)

**Comment by twisted-trac** [comment]
|[<img alt="staceysern's avatar" src="https://avatars.githubusercontent.com/u/3631162?s=50" width="50" height="50">](https://github.com/staceysern)<a name="note_4"></a>|@staceysern commented|
|-|-|

(In [40018]) Add topfile

refs [#6739](https://github.com/twisted/twisted/issues/6739)

**Comment by twisted-trac** [comment]
|[<img alt="staceysern's avatar" src="https://avatars.githubusercontent.com/u/3631162?s=50" width="50" height="50">](https://github.com/staceysern)<a name="note_5"></a>|@staceysern commented|
|-|-|

(In [40019]) Revise topfile

refs [#6739](https://github.com/twisted/twisted/issues/6739)

**Comment by twisted-trac** [comment]
|[<img alt="staceysern's avatar" src="https://avatars.githubusercontent.com/u/3631162?s=50" width="50" height="50">](https://github.com/staceysern)|@staceysern set owner to @hawkowl|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_9"></a>|@hawkowl set owner to @staceysern|
|-|-|

Hi shira, thanks for working on this! Sorry for the few days delay!

1. Queue.getPath() (line 432) has no @return or @rtype.

2. Subjectively: I don't like the second part of relaymanager.py's docstring. It does seem to work as is, so it's fine to leave it if you disagree, but it did take reading it a few times for me to actually get what it was talking about.

3. Subjectively: In `SmartHostSMTPRelayingManager.__init__`, I don't particularly like the wording of "Default values are meant for a small box with 1-5 users.", since this sounds like it has performance implications, rather than the small connections at once. Maybe add that it's the connection limits that would support 1-5 users best?

Please address [#11338](https://github.com/twisted/twisted/issues/11338) and resubmit, with anything for 2 and 3 if you agree. Thanks again.

**Comment by twisted-trac** [comment]
|[<img alt="staceysern's avatar" src="https://avatars.githubusercontent.com/u/3631162?s=50" width="50" height="50">](https://github.com/staceysern)<a name="note_10"></a>|@staceysern commented|
|-|-|

(In [40251]) Incorporate code review comments

refs [#6739](https://github.com/twisted/twisted/issues/6739)

**Comment by twisted-trac** [comment]
|[<img alt="staceysern's avatar" src="https://avatars.githubusercontent.com/u/3631162?s=50" width="50" height="50">](https://github.com/staceysern)<a name="note_11"></a>|@staceysern commented|
|-|-|

(In [40253]) Fix trailing whitespace

refs [#6739](https://github.com/twisted/twisted/issues/6739)

**Comment by twisted-trac** [comment]
|[<img alt="staceysern's avatar" src="https://avatars.githubusercontent.com/u/3631162?s=50" width="50" height="50">](https://github.com/staceysern)<a name="note_12"></a>|@staceysern removed owner|
|-|-|

Fixed item 1.  Revised comments on 2 and 3.  Please let me know if you think they are better.

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_13"></a>|@thijstriemstra set owner to @staceysern|
|-|-|

There are also quite a few occurences of wrong spacing:

```
return {'directory' : self.directory}
```

vs

```
return {'directory': self.directory}
```

and:

```
message+'-H'
```

vs 
```
message + '-H'
```

I understand this isn't a docstring issue but its about whitespace. Can you fix that whitespace as well to make the code more readable?

**Comment by twisted-trac** [comment]
|[<img alt="staceysern's avatar" src="https://avatars.githubusercontent.com/u/3631162?s=50" width="50" height="50">](https://github.com/staceysern)<a name="note_14"></a>|@staceysern commented|
|-|-|

(In [40518]) Fix additional whitespace issues and long lines

refs [#6739](https://github.com/twisted/twisted/issues/6739)

**Comment by twisted-trac** [comment]
|[<img alt="staceysern's avatar" src="https://avatars.githubusercontent.com/u/3631162?s=50" width="50" height="50">](https://github.com/staceysern)<a name="note_15"></a>|@staceysern commented|
|-|-|

(In [40519]) Revert changes to web/_newclient.py

refs [#6739](https://github.com/twisted/twisted/issues/6739)

**Comment by twisted-trac** [comment]
|[<img alt="staceysern's avatar" src="https://avatars.githubusercontent.com/u/3631162?s=50" width="50" height="50">](https://github.com/staceysern)<a name="note_16"></a>|@staceysern commented|
|-|-|

(In [40520]) Remove trailing whitespace

refs [#6739](https://github.com/twisted/twisted/issues/6739)

**Comment by twisted-trac** [comment]
|[<img alt="staceysern's avatar" src="https://avatars.githubusercontent.com/u/3631162?s=50" width="50" height="50">](https://github.com/staceysern)|@staceysern removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="hynek's avatar" src="https://avatars.githubusercontent.com/u/41240?s=50" width="50" height="50">](https://github.com/hynek)<a name="note_18"></a>|@hynek set owner to @staceysern|
|-|-|

1.[I’ve forced a build](https://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/relay-doc-6739), everything looks happy,  no pydoctor errors within `relay(manager)?.py`.
 1. relay.py
  1. 39: maybe “a … provider” would be better?
 1. relaymanager.py
  1.  `@type secret: See L{__init__}` doesn’t work, I think you should use `@ivar secret: see L{__init__}` (occurs several times).
  1. 448: I suppose it’s a list of 2-tuples? I think that should be explicitly noted as in the other cases.
  1. 640: I find this 1–5 users somewhat problematic since it depends on the machine and the amount of mail those 1–5 users receive. Maybe just call it “a personal low-volume smart host”? This one is optional and should probably be discussed with someone with more insight.
  1. 939: would you consider “which successfully fires with L{Record_MX}” really part of the type? Do we this elsewhere too? I’d just keep it in the @return.
  1. 963: `provider*s*` seems more apt here I think
  1. 982: same
  1. 1068: there is both “MX record” and “mail exchange record” in one sentence. I’d prefer if it were consistent.

Since it has been already proof-read by a native speaker and those points are rather minor, you can merge after fixing 3.1 + 3.2 + whatever you consider fix-worthy as a native speaker yourself.

**Comment by twisted-trac** [comment]
|[<img alt="staceysern's avatar" src="https://avatars.githubusercontent.com/u/3631162?s=50" width="50" height="50">](https://github.com/staceysern)<a name="note_19"></a>|@staceysern commented|
|-|-|

(In [41214]) Incorporate code review comments and update tuple and dict notation.

refs [#6739](https://github.com/twisted/twisted/issues/6739)

**Comment by twisted-trac** [comment]
|[<img alt="staceysern's avatar" src="https://avatars.githubusercontent.com/u/3631162?s=50" width="50" height="50">](https://github.com/staceysern)<a name="note_20"></a>|@staceysern commented|
|-|-|

3.1 Great catch.
3.2 It's a list of two bytes so I changed it to "L{list} of two L{bytes}"  so it doesn't look like tuple notation.
3.3 I changed it to "low-volume smart host".
3.4 I think that when the return type of a deferred is meant to be used (as opposed to using the deferred to just wait for something to happen), it should be part of the type.  The return value of getMX is really RecordMX.  It's just not immediately available so the deferred is returned.  If that part was only included in the text of the @return, important detailed information would be missing.
3.5&3.6 I agree that in terms of English, plural seems better here.  For better or worse, the style that I used throughout uses the type in singular as in L{list} of L{int}.  Among other reasons, the s would have to go outside of the braces which looks a little funny and may render in different size type in the API documentation.
3.7 Done.

I also changed the tuple and dict notation in response to comments on other mail documentation tickets.  Because of this change I am putting the ticket back into review but if they are considered minor enough I can merge.

**Comment by twisted-trac** [comment]
|[<img alt="hynek's avatar" src="https://avatars.githubusercontent.com/u/41240?s=50" width="50" height="50">](https://github.com/hynek)<a name="note_21"></a>|@hynek commented|
|-|-|

I don’t see the low-volume fix, otherwise this looks fine, please merge (after fixing the 1–5/low-volume thing if you are inclined to do so).

**Comment by twisted-trac** [comment]
|[<img alt="staceysern's avatar" src="https://avatars.githubusercontent.com/u/3631162?s=50" width="50" height="50">](https://github.com/staceysern)<a name="note_22"></a>|@staceysern commented|
|-|-|

(In [41215]) Change docstring wording.

refs [#6739](https://github.com/twisted/twisted/issues/6739)

**Comment by twisted-trac** [comment]
|[<img alt="staceysern's avatar" src="https://avatars.githubusercontent.com/u/3631162?s=50" width="50" height="50">](https://github.com/staceysern)<a name="note_23"></a>|@staceysern set status to `closed`|
|-|-|

(In [41216]) Merge <relay-doc-6739: Fix docstrings for relay.py and relaymanager.py

Author: shira
Reviewer: hawkowl, thijs, hynek
Fixes: [#6739](https://github.com/twisted/twisted/issues/6739)
